lib/src/library_analyzer.dart

```diff
@@ -0,0 +1,184 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/file_system/file_system.dart' hide File;
+import 'package:analyzer/file_system/physical_file_system.dart';
+import 'package:analyzer/source/package_map_resolver.dart';
+import 'package:analyzer/source/pub_package_map_provider.dart';
+import 'package:analyzer/src/dart/sdk/sdk.dart' show FolderBasedDartSdk;
+import 'package:analyzer/src/generated/engine.dart';
+import 'package:analyzer/src/generated/java_io.dart';
+import 'package:analyzer/src/generated/sdk.dart' show DartSdk;
+import 'package:analyzer/src/generated/source.dart';
+import 'package:analyzer/src/generated/source_io.dart';
+import 'package:cli_util/cli_util.dart' as cli;
+import 'package:path/path.dart' as p;
+
+import 'utils.dart';
+
+class LibraryScanner {
+  final String _package;
+  final String _projectPath;
+  final Map<String, List<String>> _cachedLibs = {};
+
+  AnalysisContext _context;
+  UriResolver _packageResolver;
+
+  LibraryScanner(this._package, this._projectPath) {
+    _init();
+  }
+
+  Future<Map<String, List<String>>> scanDirectLibs() =>
+      toSortedMap(_scanPackage(_package, _projectPath));
+
+  Future<Map<String, List<String>>> scanTransitiveLibs() async {
+    Map<String, List<String>> results = {};
+    Map<String, List<String>> direct =
+        await _scanPackage(_package, _projectPath);
+    for (String key in direct.keys) {
+      Set<String> processed = new Set();
+      Set<String> todo = new Set.from(direct[key]);
+      while (todo.isNotEmpty) {
+        String lib = todo.first;
+        todo.remove(lib);
+        if (processed.contains(lib)) continue;
+        processed.add(lib);
+        if (lib.startsWith('dart:')) {
+          // nothing to do
+        } else if (_cachedLibs.containsKey(lib)) {
+          todo.addAll(_cachedLibs[lib]);
+        } else if (lib.startsWith('package:')) {
+          todo.addAll(await _scanUri(lib));
+        }
+      }
+
+      results[key] = processed.toList()..sort();
+    }
+    return toSortedMap(results);
+  }
+
+  void _init() {
+    // TODO: fail more clearly if this...fails
+    var sdkPath = cli.getSdkDir().path;
+
+    JavaSystemIO.setProperty("com.google.dart.sdk", sdkPath);
+    var resourceProvider = PhysicalResourceProvider.INSTANCE;
+    DartSdk sdk = new FolderBasedDartSdk(
+        resourceProvider, resourceProvider.getFolder(sdkPath));
+
+    _packageResolver = _getPackageResolver(sdk);
```

Comment:
points for making a private ctor + public factory and having all fields private - but not a huge deal